src/main/java/de/tum/in/www1/artemis/service/connectors/gitlabci/GitLabCIService.java

```diff
@@ -0,0 +1,410 @@
+            gitlab.getPipelineApi().deletePipelineTrigger(repositoryPath, trigger.getId());
+        }
+        catch (GitLabApiException e) {
+            throw new GitLabCIException("Error triggering the build for " + repositoryPath, e);
+        }
+    }
+
+    @Override
+    public void deleteProject(String projectKey) {
+        log.error("Unsupported action: GitLabCIService.deleteBuildPlan()");
+        log.error("Please refer to the repository for deleting the project. The build plan can not be deleted separately.");
+    }
+
+    @Override
+    public void deleteBuildPlan(String projectKey, String buildPlanId) {
+        log.error("Unsupported action: GitLabCIService.deleteBuildPlan()");
+        log.error("Please refer to the repository for deleting the project. The build plan can not be deleted separately.");
+    }
+
+    @Override
+    public String getPlanKey(Object requestBody) throws ContinuousIntegrationException {
+        TestResultsDTO dto = TestResultsDTO.convert(requestBody);
+        return dto.getFullName();
+    }
+
+    @Override
+    public AbstractBuildResultNotificationDTO convertBuildResult(Object requestBody) {
+        return TestResultsDTO.convert(requestBody);
+    }
+
+    @Override
+    public BuildStatus getBuildStatus(ProgrammingExerciseParticipation participation) {
+        try {
+            final Optional<Pipeline> optionalPipeline = getLatestPipeline(participation);
+            if (optionalPipeline.isEmpty()) {
+                return BuildStatus.INACTIVE;
+            }
+            return convertPipelineStatusToBuildStatus(optionalPipeline.get().getStatus());
+        }
+        catch (GitLabApiException e) {
+            return BuildStatus.INACTIVE;
+        }
+    }
+
+    private BuildStatus convertPipelineStatusToBuildStatus(PipelineStatus status) {
+        if (status.equals(PipelineStatus.CREATED) || status.equals(PipelineStatus.WAITING_FOR_RESOURCE) || status.equals(PipelineStatus.PREPARING)
+                || status.equals(PipelineStatus.PENDING)) {
+            return BuildStatus.QUEUED;
+        }
+        else if (status.equals(PipelineStatus.RUNNING)) {
+            return BuildStatus.BUILDING;
+        }
+        else {
+            return BuildStatus.INACTIVE;
+        }
+    }
+
+    private Optional<Pipeline> getLatestPipeline(final ProgrammingExerciseParticipation participation) throws GitLabApiException {
+        final String repositoryPath = getRepositoryPath(participation.getVcsRepositoryUrl());
+        final Optional<String> commitHash = participation.findLatestSubmission().map(ProgrammingSubmission.class::cast).map(ProgrammingSubmission::getCommitHash);
+        if (commitHash.isEmpty()) {
+            return Optional.empty();
+        }
+
+        return gitlab.getPipelineApi().getPipelinesStream(repositoryPath, new PipelineFilter().withSha(commitHash.get())).max(Comparator.comparing(Pipeline::getFinishedAt));
```

Comment:
Thanks, using the filter looks even better than the suggestion. Splitting the method into smaller helper ones looks great, too!

```suggestion
        return gitlab.getPipelineApi().getPipelinesStream(repositoryPath, new PipelineFilter().withSha(commitHash.get())).max(Comparator.comparing(Pipeline::getId));
```
I think finishedAt is not the best way to sort them: e.g., there are multiple runners and one is under heavy load, another not. Then it could happen that a pipeline that was triggered first finishes later.

GitLab uses incrementing IDs so that should work, or alternatively `getStartedAt`.